### ISSUE TITLE
Incomplete "oslibs"

### ISSUE THREAD

**Issue by vanrein** [arpa2/mkhere#30]
At the very least the `/mkhere/pacman.sh` target does not list its dependent libraries in the `oslibs` command.

The binaries are at least

```
shell$ /mkhere/pacman.sh list | grep bin
usr/bin/makepkg-template
usr/bin/repo-add
usr/bin/repo-elephant
usr/bin/pacman-db-upgrade
usr/bin/pacman-key
usr/bin/testpkg
usr/bin/pacman-conf
usr/bin/pacman
usr/bin/repo-remove
usr/bin/vercmp
usr/bin/makepkg
```

The dependencies on libraries includes

```
shell$ LD_LIBRARY_PATH=/tree/pacman-5.2.1/usr/lib ldd /tree/pacman-5.2.1/usr/bin/pacman
        linux-vdso.so.1 (0x00007ffd4f3d9000)
        libalpm.so.12 => /tree/pacman-5.2.1/usr/lib/libalpm.so.12 (0x00007f99b82b5000)
        libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f99b7fc5000)
        libarchive.so.13 => /lib/x86_64-linux-gnu/libarchive.so.13 (0x00007f99b7f0c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f99b7d89000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f99b7bc8000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f99b7bc3000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f99b7ba0000)
        libnettle.so.6 => /lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f99b7b68000)
        libacl.so.1 => /lib/x86_64-linux-gnu/libacl.so.1 (0x00007f99b7b5d000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f99b7b35000)
        libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f99b7a95000)
        liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1 (0x00007f99b7a76000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f99b7a61000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f99b7843000)
        libxml2.so.2 => /lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f99b769a000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f99b830e000)
        libattr.so.1 => /lib/x86_64-linux-gnu/libattr.so.1 (0x00007f99b7692000)
        libicui18n.so.63 => /lib/x86_64-linux-gnu/libicui18n.so.63 (0x00007f99b73b7000)
        libicuuc.so.63 => /lib/x86_64-linux-gnu/libicuuc.so.63 (0x00007f99b71e6000)
        libicudata.so.63 => /lib/x86_64-linux-gnu/libicudata.so.63 (0x00007f99b57f6000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f99b5672000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f99b5658000)
```

But the only ones found are

```
shell$ /mkhere/pacman.sh oslibs
lib64/ld-linux-x86-64.so.2
usr/lib/x86_64-linux-gnu/ld-2.28.so
lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
usr/lib/x86_64-linux-gnu/ld-2.28.so
usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
usr/lib/x86_64-linux-gnu/ld-2.28.so
```

A possible cause is that Pacman relies on a library which has most of the dependencies, and perhaps this is not properly processed in `oslibs` yet?

```
bash$ ldd /tree/pacman-5.2.1/usr/lib/libalpm.so
        linux-vdso.so.1 (0x00007ffd1b3e4000)
        libarchive.so.13 => /lib/x86_64-linux-gnu/libarchive.so.13 (0x00007f72d0e0d000)
        libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f72d0b24000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f72d09a1000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f72d07e0000)
        libnettle.so.6 => /lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f72d07a8000)
        libacl.so.1 => /lib/x86_64-linux-gnu/libacl.so.1 (0x00007f72d079d000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f72d0773000)
        libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f72d06d3000)
        liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1 (0x00007f72d06b4000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f72d06a1000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f72d0483000)
        libxml2.so.2 => /lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f72d02da000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f72d02d3000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f72d02b2000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f72d0f01000)
        libattr.so.1 => /lib/x86_64-linux-gnu/libattr.so.1 (0x00007f72d02aa000)
        libicui18n.so.63 => /lib/x86_64-linux-gnu/libicui18n.so.63 (0x00007f72cffcf000)
        libicuuc.so.63 => /lib/x86_64-linux-gnu/libicuuc.so.63 (0x00007f72cfe00000)
        libicudata.so.63 => /lib/x86_64-linux-gnu/libicudata.so.63 (0x00007f72ce40e000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f72ce28a000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f72ce270000)
```

**Comment by vanrein** [comment]
Recomputing by hand, "Second, find linked libraries (dropping system-specific ones)" yields a `"$REFDIR"` value `\/usr\/lib` which is the directory of `/usr/lib/libalpm.so`.  This is prefixed to each of the libraries that lack a leading `/` in their names.

**That is silly.**  Why would the library be in the same directory?  Not just because the name is mentioned.

The problem seems to be looking at the part before `=>` and not the path that follows it.

**Comment by vanrein** [comment]
Inserted `-e 's/^[^ ]* => *//'` before `-e 's/^[^ ]* => *//'` and indeed we find

```
shell$ ldd "/tree/pacman-5.2.1/$FILE" | sed -e 's/^[ \t]*//' -e 's/^[^ ]* => //' -e 's/ .*//' -e '/^linux-vdso\.so.*/d' -e 's/^\([^/]*\)$/'"$REFDIR"'\/\1/'
/lib/x86_64-linux-gnu/libarchive.so.13
/lib/x86_64-linux-gnu/libcrypto.so.1.1
/lib/x86_64-linux-gnu/libm.so.6
/lib/x86_64-linux-gnu/libc.so.6
/lib/x86_64-linux-gnu/libnettle.so.6
/lib/x86_64-linux-gnu/libacl.so.1
/lib/x86_64-linux-gnu/liblzma.so.5
/lib/x86_64-linux-gnu/libzstd.so.1
/lib/x86_64-linux-gnu/liblz4.so.1
/lib/x86_64-linux-gnu/libbz2.so.1.0
/lib/x86_64-linux-gnu/libz.so.1
/lib/x86_64-linux-gnu/libxml2.so.2
/lib/x86_64-linux-gnu/libdl.so.2
/lib/x86_64-linux-gnu/libpthread.so.0
/lib64/ld-linux-x86-64.so.2
/lib/x86_64-linux-gnu/libattr.so.1
/lib/x86_64-linux-gnu/libicui18n.so.63
/lib/x86_64-linux-gnu/libicuuc.so.63
/lib/x86_64-linux-gnu/libicudata.so.63
/lib/x86_64-linux-gnu/libstdc++.so.6
/lib/x86_64-linux-gnu/libgcc_s.so.1
```
instead of

```
shell$ ldd "/tree/pacman-5.2.1/$FILE" | sed -e 's/^[ \t]*//' -e 's/ .*//' -e '/^linux-vdso\.so.*/d' -e 's/^\([^/]*\)$/'"$REFDIR"'\/\1/'
/usr/lib/libarchive.so.13
/usr/lib/libcrypto.so.1.1
/usr/lib/libm.so.6
/usr/lib/libc.so.6
/usr/lib/libnettle.so.6
/usr/lib/libacl.so.1
/usr/lib/liblzma.so.5
/usr/lib/libzstd.so.1
/usr/lib/liblz4.so.1
/usr/lib/libbz2.so.1.0
/usr/lib/libz.so.1
/usr/lib/libxml2.so.2
/usr/lib/libdl.so.2
/usr/lib/libpthread.so.0
/lib64/ld-linux-x86-64.so.2
/usr/lib/libattr.so.1
/usr/lib/libicui18n.so.63
/usr/lib/libicuuc.so.63
/usr/lib/libicudata.so.63
/usr/lib/libstdc++.so.6
/usr/lib/libgcc_s.so.1
```

**Comment by vanrein** [comment]
Wiped the `/build/pacman-5.2.1/mkhere-oslibs.txt` and got a much longer list, including traversal of links.

**Comment by vanrein** [comment]
Resolved with 4feeff7250a7c8e22b4c3eae88054b5c8906d605 in branch `master`.
